Add EVPN support to RouteAdvertisements controller

kyrtapz opened the pull request:
Based on: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5801

Generate FRR configuration for networks with EVPN configured:
- Activate EVPN on global router with advertise-all-vni
- MAC-VRF: configure route targets if specified
- IP-VRF: create VRF-to-VNI mappings and per-VRF BGP routers with
  unicast advertisement and route targets
- Create routers for IP-VRF networks if not already selected
Uses raw FRR config until frr-k8s provides typed EVPN API.

Note: I've opted not to explicitly set the defaults for RD/RT and let FRR handle it. This ensures we rely on the default behavior. 

Remaining bit is to figure out if we are ok with the possible repetition/duplication when multiple frrconfigs are selected by an RA. 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added EVPN support for route advertisements, including MAC‑VRF and IP‑VRF vrf/VNI handling and generated EVPN FRR configuration.

* **Improvements**
  * Deterministic ordering and stronger validation for EVPN networks (router requirements, consistency checks).
  * Improved diagnostic message when EVPN transport is requested but the feature is disabled.

* **Tests**
  * Added comprehensive tests covering MAC‑VRF, IP‑VRF, dual‑stack, and combined EVPN scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kyrtapz commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -720,11 +782,49 @@ func (c *Controller) generateFRRConfiguration(
+			}
+		}
+
+		if globalRouterASN == 0 {
```
This check enforces that every frrconfig matched by an RA for an EVPN network must contain a global router. I think this is acceptable—it keeps the logic simple, and the potential repetition shouldn't be a problem for frr-k8s. I’d like to hear what others think.

If we were to loosen this constraint, we would need to perform the check outside of generateFRRConfiguration and only configure EVPN bits for frrconfigs that actually contain a global router.

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/evpn_rawconfig.go:
```diff
@@ -0,0 +1,134 @@
+package routeadvertisements
+
+// TODO: Delete this file when frr-k8s adds typed API support for EVPN.
```
TODO makes more sense to me where `generateEVPNRawConfig` is called

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -385,6 +411,40 @@ func (c *Controller) generateFRRConfigurations(ra *ratypes.RouteAdvertisements)
+		}
+		if (len(selectedNetworks.macVRFConfigs) > 0 || len(selectedNetworks.ipVRFConfigs) > 0) && ra.Spec.TargetVRF == "" {
+			return nil, nil, fmt.Errorf("%w: advertising networks with EVPN configuration to default target VRF is not supported", errConfig)
+		}
```
I would be more restrictive with this check. I would check that TargetVRF is either `auto` or equal to the CUDN vrf. Otherwise you end up in a situation were you are configuring prefixes in the unicast address family on one router but configuring the IPVRF in a different router (since you are not using the target VRF in line 440).

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -720,11 +782,49 @@ func (c *Controller) generateFRRConfiguration(
+			ASN:      globalRouterASN,
+			VRF:      cfg.VRFName,
+			Prefixes: selectedNetworks.hostNetworkSubnets[cfg.NetworkName],
+		})
```
I guess I have a similar comment as before, to be able to find the router defined in any of the configs. With raw config this is not a problem, but we will need this with the formal API.

You could just look into `matchedNetworks` at the end of the process and generate an extra configuration for the unmatched ones.

I guess I am fine with the current approach but we will need to do something later.

BTW, aren't we triggering this error in this scenario?
```
		// check that we matched all the selected networks on 'auto'
		if ra.Spec.TargetVRF == "auto" && !matchedNetworks.HasAll(selectedNetworks.networks...) {
			return nil, nil, fmt.Errorf("%w: selected FRRConfigurations for node %q don't match all selected networks with target VRF 'auto'", errConfig, nodeName)
		}
```

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -720,11 +782,49 @@ func (c *Controller) generateFRRConfiguration(
+		klog.Infof("Creating router for EVPN network %q VRF %q with ASN=%d, prefixes=%v",
+			cfg.NetworkName, cfg.VRFName, globalRouterASN, selectedNetworks.hostNetworkSubnets[cfg.NetworkName])
+		routers = append(routers, frrtypes.Router{
+			ASN:      globalRouterASN,
```
hmm I need to check what's the effect of this ASN, makes sense that we would set it to the underlay one but what would actually happen if it was different?

kyrtapz replied:
Yeah, I think it is a reasonable default and if a user provides the vrf router we should use it's asn instead.

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -720,11 +782,49 @@ func (c *Controller) generateFRRConfiguration(
+	if len(routers) == 0 && len(selectedNetworks.macVRFConfigs) == 0 {
+		// we ended up with no routers and there is no macVRF to configure, bail out
+		return nil, nil
+	}
```
wouldn't we prematurely bail out in an scenario where we just want to apply the global evpn configuration?

kyrtapz replied:
We wouldn't, it's an and condition.

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -385,6 +417,41 @@ func (c *Controller) generateFRRConfigurations(ra *ratypes.RouteAdvertisements)
+				HasIPv6:     hasIPv6,
+			})
+		}
+		if (len(selectedNetworks.macVRFConfigs) > 0 || len(selectedNetworks.ipVRFConfigs) > 0) && (ra.Spec.TargetVRF != "auto" && ra.Spec.TargetVRF != vrf) {
```
I usually like to break out this long conditions to some more readable form like
```
hasEVPNConfig := len(selectedNetworks.macVRFConfigs) > 0 || len(selectedNetworks.ipVRFConfigs) > 0

if hasEVPNConfig && ra.Spec.TargetVRF != "auto" && ra.Spec.TargetVRF != vrf {
```

Subjective style but sharing with you in case you will take it

jcaamano commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -577,6 +661,10 @@ func (c *Controller) generateFRRConfigurations(ra *ratypes.RouteAdvertisements)
 			return nil, nil, fmt.Errorf("%w: selected FRRConfigurations for node %q don't match all selected networks with target VRF 'auto'", errConfig, nodeName)
 		}
+		// check that we found a global router if MAC-VRF EVPN is configured.
+		if len(selectedNetworks.macVRFConfigs) > 0 && !matchedNetworks.Has(evpnGlobalRouterKey) {
```
What is the practical difference of `matchedNetworks.Has(evpnGlobalRouterKey)` vs `frrRouterVRFs.Has("")`? If there is none, we could consolidate to one of the approaches. If there is any, then are we sure we need to different interpretations of what the global router is?

kyrtapz commented:
Rebase: https://github.com/ovn-kubernetes/ovn-kubernetes/compare/3b4ba3b4bdf107f76954bccf6886ac98af1348bf..ebb0fb8814393be92313e6daa69d942fa183056e

kyrtapz commented:
Addressing review comments: https://github.com/ovn-kubernetes/ovn-kubernetes/compare/ebb0fb8814393be92313e6daa69d942fa183056e..6c89de7b90972aeada2470f35344494c47ac3ec5

jcaamano reviewed (commented):
I have one remaining concern.

There is an scenario where we might be selecting a network with just an IP-VRF and FRRConfiguration with a router on that VRF but no router on default VRF. The result is no underlay EVPN configuration generated, and if that is also not generated through a different RA, we end up with a configuration that doesn't work.

So I am guessing if we need to actually require a router on the default VRF being selected. While there are other approaches this might be the simplest

kyrtapz commented:
hit a known flake: https://github.com/ovn-kubernetes/ovn-kubernetes/issues/4221 
https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/21512455093/job/61983553800?pr=5807
